### ISSUE TITLE
Cannot reproduce paper results with released RoboTwin-MeM checkpoint (0/20 across 3 tasks vs paper's 48–90%, after fixing several setup issues)

### ISSUE THREAD

**Issue by jiang-tianyi666** [InternRobotics/EventVLA#3]
Thanks for releasing the code and checkpoints! We spent considerable effort evaluating the released RoboTwin-MeM checkpoint and would like to report our findings and ask a few questions.

## Setup issues we had to fix first

(sharing in case they help others)

1. **Assets**: as in #2, the download script pulls RoboTwin 2.0 base objects whose indices collide with RoboTwin-MeM objects (e.g. `009_kettle` vs `009_toycar`, `010_pen` vs `010_mouse`). Fixed by overlaying the full assets from the HF dataset repo.
2. **Embodiment**: the training data metadata (`lerobot_2.1/*/meta/info.json`) says `robot_type: "aloha"`, so we evaluate with `embodiment: [aloha-agilex]`. Note `assets/embodiments/aloha-agilex/curobo_left.yml` and `curobo_right.yml` contain hardcoded absolute paths (`/mnt/workspace/yangganlin/code/RMBench/...`) that need manual fixing.
3. **Checkpoint config**: the released `config.yaml` has `framework.name: QwenOFT`, which `build_framework` rejects; we changed it to `EventVLA`. The weights then load with **zero missing/unexpected/shape-mismatched keys**.
4. **Websocket**: client `ping_interval=20` drops the connection when one inference exceeds 20s; we set it to `None`.

## Results

After these fixes the policy behaves qualitatively correctly: the ALOHA arms reach and press buttons, and keyframe events fire (conf≈1.0) exactly at press moments, with the scene matching the training videos. However:

| Task | Paper (Table 2) | Ours (seeds 100000+, `demo_clean`, `instruction_type: unseen`, step limit 1000) |
|---|---|---|
| press_button_keyframe | 48% | **0/10** |
| put_back_block_hard | 62% | **0/5** |
| pick_objects_in_order | 90% | **0/5** |

All runs are clean (no crashes/disconnects). A typical failure: in press_button_keyframe the policy presses 2–5 times and then idles until timeout, i.e. it fails the memory-dependent counting — even though keyframes are being committed.

## Questions

1. Is `RoboTwin-MeM/final_model/pytorch_model.pt` the exact model used for Table 2?
2. Could you share the exact evaluation configuration (task_config yaml, instruction_type, step limits, seed range)?
3. Is there any known issue where committed keyframes fail to be injected into the VLM at inference in the released code path (`examples/RoboTwin-Mem/eval_files/`)?

Happy to provide full logs. Thanks!

**Comment by yelone7** [comment]
I guess it’s because you didn’t modify the camera FOV during evaluation. Based on calculations from the dataset, the actual FOV for the cover_blocks_hard task is 50.

**Comment by zhangzhengtu** [comment]
Hi @jiang-tianyi666, thank you very much for the detailed reproduction report. We also really appreciate that you listed the changes related to assets, embodiment, checkpoint config, websocket, and other details. These are helpful not only for other users trying to reproduce the results, but also for us to further improve the documentation and configuration.

Sorry for the delayed response. We have been occupied with other work recently and could not follow up earlier. To better understand why the reproduced results differ from those reported in the paper, could you please provide some additional information if possible? In particular, it would be very helpful if you could share several rollout / inference videos, especially the failure cases for press_button_keyframe.

These videos would help us determine whether the issue comes from environment configuration, action execution, memory writing / injection, success judgment, or other factors. Once we have these materials, we will compare them with our evaluation setup and try to identify the cause as soon as possible.

Hi @yelone7, thank you for the follow-up discussion. May I ask whether you have successfully reproduced the RoboTwin-MeM results on your side?

If you have successfully reproduced them, could you please share the key changes you made, such as the evaluation command, config files, seed settings, and whether any additional code or configuration patches were applied? This information would be very helpful for us and for other users who are trying to reproduce the results.

If you have not fully reproduced the results yet, feel free to share your current logs or rollout videos as well. We would be happy to compare the failure modes and see whether they are consistent with the issue reported above.

**Comment by jiang-tianyi666** [comment]
Update: the original "0/20 systemic failure" framing is no longer accurate.

**Setup fixes I applied**
- run evaluations serially with one policy server per task (EventVLA keeps server-side runtime KEM memory);
- pass `ACTION_MODE="'abs'"`, otherwise RoboTwin's `eval(value)` parses `abs` as Python's builtin;
- use the released checkpoint with the QwenOFT/EventVLA framework-name compatibility fix;
- set `eval_video_log: true` and patched only the video writer to fall back to `head_camera.rgb` when `third_view_rgb` is absent.

**Camera FOV (following @yelone7)** — I reverse-computed the intrinsics from the released HDF5. For `put_back_block_hard` and `rearrange_blocks_hard`, head/left/right camera = fovy 50, which matches the current `LargeView` config. The policy input only uses `[head, left, right]` (`front_camera` is fovy 37 in the data but is not read by the interface), so these two tasks do **not** appear to fail from FOV OOD.

**10-episode video diagnostics** (all `rc=0`, no traceback / missing asset / websocket timeout):
- `press_button_keyframe`: **8/10** (and 58/97 ≈ 60% on a longer run) — at or above the paper's 48%, so the pipeline itself looks sound.
- `put_back_block_hard`: **0/10**, max reward is **exactly 0.86 in every episode** (110/110 across all my runs, never 0.94+), all episodes run to the full 500/500 steps.
- `rearrange_blocks_hard`: **0/10**, max reward 0.6–0.9, all run to the full 700/700 steps.

The `put_back` / `rearrange` videos are not initial visual OOD: pads/blocks are visible from the start and the arm does not aim off-scene. `put_back` looks like it reliably completes the first block's return but never the "both blocks restored + final press" stage; `rearrange` looks like an ordering / terminal-condition failure.

Given that `press_button` reproduces at/above paper while `put_back` collapses 62% → 0%, I suspect something **task-specific** rather than a general memory limitation, and I can't yet tell whether the policy fails to complete these tasks or whether the success check isn't triggering. Two questions:

1. Do the Table 2 numbers for put_back / rearrange correspond to the **`_hard`** variants (`put_back_block_hard` / `rearrange_blocks_hard`) that I'm running?
2. For `put_back`, does a max reward pinned at exactly **0.86** (first block back on its original outer pad, never reaching the "both blocks restored" stage) match your internal runs, or could the success/reward check be sensitive to the RoboTwin version? Could you share your exact eval config for these two tasks (seed range, step limit, `ckpt_setting`)?

**Representative failure videos** (download links):
- [put_back_block_hard — episode 0 (fail, reward capped at 0.86)](https://github.com/jiang-tianyi666/eventvla-repro-videos/releases/download/v1/put_back_ep0_fail.mp4)
- [rearrange_blocks_hard — episode 0 (fail)](https://github.com/jiang-tianyi666/eventvla-repro-videos/releases/download/v1/rearrange_ep0_fail.mp4)
- [press_button_keyframe — episode 0 (fail)](https://github.com/jiang-tianyi666/eventvla-repro-videos/releases/download/v1/press_button_ep0_fail.mp4)
- [press_button_keyframe — episode 5 (fail)](https://github.com/jiang-tianyi666/eventvla-repro-videos/releases/download/v1/press_button_ep5_fail.mp4)

**Comment by zhangzhengtu** [comment]
Hi @jiang-tianyi666, thank you very much for the detailed reproduction feedback, and thanks also to @yelone7 for the additional clarification about the camera FOV. Based on your updated information, we re-checked our evaluation configuration and also re-ran a batch evaluation locally.

Let me first answer two key questions:

1. The `put_back` / `rearrange` tasks in Table 2 correspond to the hard versions, i.e., `put_back_block_hard` and `rearrange_blocks_hard`.
2. `RoboTwin-MeM/final_model/steps_100000_pytorch_model.pt` is the released checkpoint used for the `pure_image_keyframe_memory` setting in Table 2. Although the old code uses `QwenOFT` as `framework.name`, it is still compatible with the current code path and can be loaded by `build_framework`.

The evaluation entry we currently use locally is:

```bash
bash examples/RoboTwin-Mem/eval_batch/run_batch_eval.sh \
  examples/RoboTwin-Mem/eval_batch/weights_8tasks_pure_image_keyframe_memory_teacher_qwenoft.sh
```

The core configuration is:

```text
TASK_CONFIG=demo_clean
INSTRUCTION_TYPE=unseen
ACTION_MODE=abs
UNNORM_KEY=new_embodiment
CKPT_SETTING=pure_image_keyframe_memory
seed=0
camera=LargeView
```

Here, `seed=0` is converted in `script/eval_policy.py` to the starting seed `100000 * (1 + seed)`, so the actual evaluation seeds start from `100000`.

For the camera setting, `demo_clean.yml` uses `LargeView`, which corresponds to the following entry in `_camera_config.yml`:

```text
LargeView:
  fovy: 50
  w: 320
  h: 240
```

This is consistent with the FOV you reverse-computed.

At the moment, I think the most suspicious discrepancy is the step limit. In your updated results, you mentioned:

```text
put_back_block_hard: full 500/500 steps
rearrange_blocks_hard: full 700/700 steps
```

However, in our current code, the step limits for the hard-version tasks are:

```text
put_back_block_hard: 1800
rearrange_blocks_hard: 1200
press_button_keyframe: 1500
pick_objects_in_order: 1500
```

The values `500` and `700` exactly match the step limits of the non-hard versions, namely `put_back_block` and `rearrange_blocks`. Therefore, if the hard tasks on your side actually ran only for 500 / 700 steps, the issue is likely caused by an inconsistent `_eval_step_limit.yml`, a task-name mapping mismatch, or another place where the step limit is overwritten.

Our local running logs also support this hypothesis. The currently written episode results are:

```text
put_back_block_hard:
  8/8 success, all with max reward 1.0
  seeds: 100000, 100001, 100003, 100004, 100006, 100007, 100008, 100010

rearrange_blocks_hard:
  9/11 success
  failed seeds: 100000, 100002

press_button_keyframe:
  11/15 success
  failed seeds: 100000, 100002, 100005, 100009

pick_objects_in_order:
  5/9 success
```

Based on the current results, `put_back_block_hard` does not show a systematic 0% failure, and `press_button_keyframe` can also be reproduced at a level higher than or close to the number reported in the paper table. Therefore, we currently do not see a general issue in the released code path where “keyframes are committed but not injected into the VLM”. The more likely cause is that some task-level evaluation configurations are not fully aligned.

**Comment by jiang-tianyi666** [comment]
Thank you @zhangzhengtu — you were exactly right, and this resolves it.

The root cause was precisely what you pointed out: my `_eval_step_limit.yml` had `500` / `700` for `put_back_block_hard` / `rearrange_blocks_hard`, which are the **non-hard** limits. The hard tasks were being truncated before the final stage. After setting your values:

- `put_back_block_hard: 1800`
- `rearrange_blocks_hard: 1200`

I re-ran (10 episodes, seeds from 100000, serial, one policy server per task, `ACTION_MODE="'abs'"`):

- `put_back_block_hard`: **10/10**, all max reward 1.0
- `rearrange_blocks_hard`: **7/10** (failed seeds 100002 / 100004 / 100007, each capped at 0.8)

These line up with your internal numbers (put_back 8/8, rearrange 9/11). So the earlier "`put_back` reward pinned at exactly 0.86 / `rearrange` stuck at 0.8–0.9, running the full step budget" was simply the hard task hitting the non-hard step limit before completion — not a memory failure and not a broken success check. FOV was already aligned (head/left/right = fovy 50).

Net: with the correct per-task step limits, I no longer see any systematic failure, and I see no evidence of a "keyframes committed but not injected into the VLM" problem in the released path — the 0% was entirely on my side from an inconsistent `_eval_step_limit.yml`. Sorry for the earlier alarming framing, and thanks a lot for the quick and precise help!

**Comment by yunhang8658** [comment]
你好，@jiang-tianyi666,@zhangzhengtu我们也在复现 EventVLA 的 RoboTwin-MeM 结果。使用官方公开的 RoboTwin-MeM/final_model/pytorch_model.pt，评测 pick_objects_in_order，配置为 demo_clean、
  instruction_type=unseen、action_mode=abs、unnorm_key=new_embodiment、ALOHA-AgileX、seed 100000–100099、step limit 1500，最终得到 49/100（49%），远低于论文 Table
  2 的 90%。我们也尝试将推理参数调整为论文报告的 NMS=8、cooldown=10、max_keyframes=5，100 个 episode 的结果仍然是 49%，没有改善。另外，我们发现论文写的是训练 80k
  steps、keyframe loss weight 0.1、max keyframes 5，但公开权重附带配置是 100k steps、loss weight 1.0、max keyframes 4。请问你们目前复现到了多少？是否拿到了论文实
  际使用的 80k checkpoint、对应配置、代码 commit 或作者提供的完整评测命令？如果有其他能够接近 90% 的修改，也希望能分享一下，谢谢
